ipa-core/src/protocol/context/mod.rs

```diff
@@ -344,6 +374,154 @@ impl<'a> Inner<'a> {
     }
 }
 
+/// Reshards the given set of shares across all shards available on the current helper. It uses the
+/// provided function to determine destination shards for each input row.
+///
+/// The resulting vector contains all shares sent to this shard by other shards and shares selected
+/// locally by `shard_picker` fn to stay on this shard.
+///
+/// Resharding opens one communication channel per peer (N^2 channels will be open in total,
+/// N per shard). Each channel stays open until the very last row is processed, then they are explicitly
+/// closed, even if nothing has been communicated between that pair.
+///
+/// ## Shard picking considerations
+/// It is expected for `shard_picker` to select shards uniformly, by either using [`prss`] or sampling
+/// random values with enough entropy. Failure to do so may lead to extra memory overhead - this
+/// function uses the conservative `1.2` coefficent to estimate the number of records per shard after
+/// resharding is completed, according to our [`calculations`] that is sufficient with 2^-60 failure
+/// probability. This is a very conservative estimate, assuming 1M events per shard and 100k shards.
+///
+/// [`calculations`]: https://docs.google.com/document/d/1vej6tYgNV3GWcldD4tl7a4Z9EeZwda3F5u7roPGArlU/
+///
+/// ## Panics
+/// When `shard_picker` returns an out-of-bounds index.
+///
+/// ## Errors
+/// If cross-shard communication fails
+pub async fn reshard<L, K, C, S>(
+    ctx: C,
+    input: L,
+    shard_picker: S,
+) -> Result<Vec<K>, crate::error::Error>
+where
+    L: IntoIterator<Item = K>,
+    L::IntoIter: ExactSizeIterator,
+    S: Fn(C, RecordId, &K) -> ShardIndex,
+    K: Message + Clone,
+    C: ShardedContext,
+{
+    let input = input.into_iter();
+    let input_len = input.len();
+
+    // We set channels capacity to be at least 1 to be able to open send channels to all peers.
+    // It is prohibited to create them if total records is not set. We also over-provision here
+    // because it is not known in advance how many records each peer receives. We could've set
+    // the channel capacity to be indeterminate, but it could be less efficient in using our most
+    // precious resource - network.
+    let ctx = ctx.set_total_records(std::cmp::max(1, input_len));
+    let my_shard = ctx.shard_id();
+
+    // Open communication channels to all shards on this helper and keep track of records sent
+    // through any of them.
+    let mut sending_ends = ctx
```

Comment:
`SendingEnd` is the type returned by `shard_send_channel`. You can think of it as "transmit handle" (in the sense of `let (tx, rx) = mpsc::channel()`.

Reply:
for every channel there is always at least one sender and one receiver. Sending end of a channel is what we give to the sender(s) and receiving end is owned by receiver(s).

To avoid confusion here, let me rename it to send_channels